src/frontend/src/expr/utils.rs

```diff
@@ -587,12 +583,174 @@ impl WatermarkAnalyzer {
+                    (Increasing, Constant) => Some(Increasing),
+                    (Constant, Increasing) | (Increasing, Increasing) => None,
+                }),
+            ExprType::Multiply | ExprType::Divide | ExprType::Modulus => self
+                .visit_binary_op(func_call.inputs())
+                .and_then(|m| match m {
+                    (Constant, Constant) => Some(Constant),
+                    // NOTE(stdrc) not meaningful to derive watermark for other situations
+                    _ => None,
+                }),
+            ExprType::AtTimeZone => {
+                self.visit_binary_op(func_call.inputs())
+                    .and_then(|m| match m {
+                        (Constant, Constant) => Some(Constant),
+                        (Increasing, Constant) => {
+                            if !(func_call.return_type() == DataType::Timestamptz
+                                && func_call.inputs()[0].return_type() == DataType::Timestamp)
+                                && func_call.inputs()[1]
+                                    .as_literal()
+                                    .and_then(|literal| literal.get_data().as_ref())
+                                    .map_or(true, |time_zone| {
+                                        !time_zone.as_utf8().eq_ignore_ascii_case("UTC")
+                                    })
+                            {
+                                None
+                            } else {
+                                Some(Increasing)
+                            }
+                        }
+                        _ => None,
+                    })
+            }
+
+            ExprType::AddWithTimeZone | ExprType::SubtractWithTimeZone => {
+                // Requires time zone and interval to be literal, at least for now.
+                let time_zone = match &func_call.inputs()[2] {
+                    ExprImpl::Literal(lit) => lit.get_data().as_ref().map(|s| s.as_utf8()),
+                    _ => return None,
+                };
+                let interval = match &func_call.inputs()[1] {
+                    ExprImpl::Literal(lit) => lit.get_data().as_ref().map(|s| s.as_interval()),
+                    _ => return None,
+                };
+                // null zone or null interval is treated same as const `interval '1' second`, to be
+                // consistent with other match arms.
+                let zone_without_dst = time_zone.map_or(true, |s| s.eq_ignore_ascii_case("UTC"));
+                let quantitative_only = interval.map_or(true, |v| {
+                    v.months() == 0 && (v.days() == 0 || zone_without_dst)
+                });
+                self.visit_expr(&func_call.inputs()[0])
+                    .and_then(|m| match (m, quantitative_only) {
+                        (Constant, _) => Some(Constant),
+                        (Increasing, true) => Some(Increasing),
+                        (_, false) => None,
```

Comment:
```suggestion
                        (Increasing, false) => None,
```

`(_, false)` seems likely to catch other cases and the compiler won't complain if we extend the enum later.